csrc/device_lower/analysis/thread_predicate.cpp

```diff
@@ -494,6 +503,212 @@ class RedundantUseAnalysis : BackwardVisitor {
+  const std::vector<IterDomain*>& root_domain_;
+  // leaf domains that are merged from root domains and parallelized by thread
+  // blocks
+  std::vector<IterDomain*> candidate_leaf_domains_;
+  // map from root domain to its concretized domain
+  std::unordered_map<IterDomain*, IterDomain*>
+      concretized_broadcast_root_domains_;
+  // map from parallel type to its write index
+  std::unordered_map<ParallelType, Val*> write_index_map_;
+
+  void setCandidateLeafDomains() {
+    for (auto ld : tv_->domain()->leaf()) {
+      const ParallelType& pt = ld->getParallelType();
+      auto merge = dynamic_cast<Merge*>(ld->definition());
+      if (isParallelTypeThread(pt) && merge) {
+        candidate_leaf_domains_.push_back(ld);
+      }
+    }
+  }
+  // For each broadcast root domain, find a concretized domain from its
+  // exact mapped domain set and check all the concretized domains are
+  // in the same IdMappingMode::EXACT group.
+  void setConcretizedBroadcastRootDomain() {
+    for (auto rd : root_domain_) {
+      if (rd->isBroadcast()) {
+        const auto& mapped_rd_exact_set = GpuLower::current()
+                                              ->caMap()
+                                              ->getIdSets(IdMappingMode::EXACT)
+                                              .getDisjointSetOf(rd);
+        for (auto mapped_rd : mapped_rd_exact_set) {
+          std::unordered_set<IterDomain*> all_cids =
+              GpuLower::current()
+                  ->concretizedBroadcastDomains()
+                  ->allConcretizedDomains(mapped_rd);
+          auto fromSameExactGroup = [&all_cids]() {
+            const auto& exact_set = GpuLower::current()
+                                        ->caMap()
+                                        ->getIdSets(IdMappingMode::EXACT)
+                                        .getDisjointSetOf(*all_cids.begin());
+            return std::all_of(
+                all_cids.begin(), all_cids.end(), [&](IterDomain* id) {
+                  return exact_set.has(id);
+                });
+          };
+          if (!all_cids.empty()) {
+            // assert that all_cids are in the same group in the
+            // IdMappingMode::EXACT graph
+            TORCH_INTERNAL_ASSERT(
+                fromSameExactGroup(),
+                "All concretized domains of a broadcast root domain should be in the same group in the IdMappingMode::EXACT graph.");
+            // make sure this broadcast root domain is concretized to the same
+            // domain otherwise, treat as not concretized becase we don't know
+            // which one to use.
+            auto iter = concretized_broadcast_root_domains_.find(rd);
+            if (iter != concretized_broadcast_root_domains_.end() &&
+                iter->second != *all_cids.begin()) {
+              concretized_broadcast_root_domains_.erase(iter);
+              // break to move to the next broadcast root domain
+              break;
```

Comment:
Can you explain why?

Reply:
changed from `TORCH_INTERNAL_ASSERT` to stop applying this optimization if not concretized to the same domain.